### PR TITLE
ports/all: Allow sensor_abort() to be called from different contexts.

### PR DESCRIPTION
**Pull request by iabdalkader**
This patch removes the duplicated `sensor_abort()` logic in the stm32 and mimxrt ports. This is done by adding a `flush_fifo` argument, which controls the flush operation, and a `in_irq` argument , which indicates whether `sensor_abort()` is called from within an IRQ context or not.

### CODE REVIEW

File: src/omv/ports/stm32/sensor.c [hunk 1]
```diff
@@ -229,13 +228,17 @@ int sensor_dcmi_config(uint32_t pixformat) {
...
-        HAL_NVIC_EnableIRQ(DMA2_Stream1_IRQn);
+        if (in_irq) {
+            HAL_DMA_Abort_IT(&DMAHandle);
+        } else {
+            HAL_DMA_Abort(&DMAHandle);
+        }
+        HAL_NVIC_DisableIRQ(DMA2_Stream1_IRQn);
         #if defined(OMV_MDMA_CHANNEL_DCMI_0)
         HAL_MDMA_Abort(&DCMI_MDMA_Handle0);
         HAL_MDMA_Abort(&DCMI_MDMA_Handle1);
```
Comment (kwagyeman) [inline src/omv/ports/stm32/sensor.c:244]:
```suggestion
        if (!in_irq) {
            HAL_MDMA_Abort(&DCMI_MDMA_Handle0);
            HAL_MDMA_Abort(&DCMI_MDMA_Handle1);
        }
```

Comment (kwagyeman) [reply src/omv/ports/stm32/sensor.c:244]: Looks good otherwise.

**Comment by kwagyeman** [comment]
@iabdalkader Changing the "HAL_NVIC_EnableIRQ(DMA2_Stream1_IRQn);" logic int the STM32 driver seems to break things.

```
import sensor, image, time

sensor.reset()
sensor.set_pixformat(sensor.RGB565)
sensor.set_framesize(sensor.QVGA)
sensor.set_framebuffers(1)
sensor.skip_frames(time = 2000)

clock = time.clock()

while(True):
    clock.tick()
    img = sensor.snapshot()
    print(clock.fps())
```

Just crashes. If I restore the normal behavior of that IRQ by doing:

```
    if (DCMI->CR & DCMI_CR_ENABLE) {
        DCMI->CR &= ~DCMI_CR_ENABLE;
        if (in_irq) {
            HAL_DMA_Abort_IT(&DMAHandle);
        } else {
            HAL_DMA_Abort(&DMAHandle);
            HAL_NVIC_EnableIRQ(DMA2_Stream1_IRQn);
        }
        #if defined(OMV_MDMA_CHANNEL_DCMI_0)
        if (!in_irq) {
            HAL_MDMA_Abort(&DCMI_MDMA_Handle0);
            HAL_MDMA_Abort(&DCMI_MDMA_Handle1);
        }
        HAL_MDMA_DeInit(&DCMI_MDMA_Handle0);
        HAL_MDMA_DeInit(&DCMI_MDMA_Handle1);
        #endif
        __HAL_DCMI_DISABLE_IT(&DCMIHandle, DCMI_IT_FRAME);
        __HAL_DCMI_CLEAR_FLAG(&DCMIHandle, DCMI_FLAG_FRAMERI);
        first_line = false;
        drop_frame = false;
        sensor.last_frame_ms = 0;
        sensor.last_frame_ms_valid = false;
    }
```

Then is works as expected. I think I'd feel more comfortable leaving the behavior as is. It is possible to change it, though... I just need to reload a lot of the driver state into my head again to workout why the failure is happening. I don't have time today though to focus on this.

**Comment by kwagyeman** [comment]
IMXRT works as expected.

File: src/omv/ports/stm32/sensor.c (hunk 1 above)
Comment (iabdalkader) [reply src/omv/ports/stm32/sensor.c:244]: That doesn't seem to have any effect either. I think it's safe to abort it unconditionally, but I can make that change if it's necessary.

**Comment by iabdalkader** [comment]
> Changing the "HAL_NVIC_EnableIRQ(DMA2_Stream1_IRQn);" logic int the STM32 driver seems to break things.
> 
> ```
> import sensor, image, time
> 
> sensor.reset()
> sensor.set_pixformat(sensor.RGB565)
> sensor.set_framesize(sensor.QVGA)
> sensor.set_framebuffers(1)
> sensor.skip_frames(time = 2000)
> 
> clock = time.clock()
> 
> while(True):
>     clock.tick()
>     img = sensor.snapshot()
>     print(clock.fps())
> ```

I think I just needed to enable the DMA IRQ before `HAL_DCMI_Start_DMA_xx`. I tested this with both OPENMV4/MT9M and OPENMV4P/OV5640,  and they both seem to work fine (with 1 frame buffer or more). Can you test again ?

File: src/omv/ports/stm32/sensor.c (hunk 1 above)
Comment (kwagyeman) [reply src/omv/ports/stm32/sensor.c:244]: The aborts technically have an infinite loop in them. So, you want to not call them in an IRQ.

Comment (iabdalkader) [reply src/omv/ports/stm32/sensor.c]: Good point, fixed.

**Comment by kwagyeman** [comment]
@iabdalkader - That fixed it. All good.

**Pull request merged**